packages/core/logging/core-logging-server-internal/src/logging_system.ts

```diff
@@ -110,6 +115,27 @@ export class LoggingSystem implements ILoggingSystem {
     }
   }
 
+  /**
+   * A mechanism for specifying a set of "global" {@link LogMeta} that we want
+   * to inject into all log entries.
+   *
+   * @example
+   * Global meta is provided with a lodash-style dot-separated path,
+   * and the value you wish to set:
+   * ```ts
+   * loggingSystem.setGlobalMeta('service.node.roles', ['ui']);
+   * // adds { service: { node: { roles: ['ui'] } } } to every log entry
```

Comment:
It would feel more intuitive to have a `LogMeta` object provided here. However, I went with this approach mainly for performance reasons: the log meta will get merged in for _every_ log entry. Deeply merging an object on every log entry could be very expensive, but using `lodash.set` to set a value at a particular path is a much cheaper operation.

The downside is that this approach is a little less intuitive, and it also means we are sacrificing the type safety of our ECS types (we can't enforce proper ECS fields are being used via `setGlobalMeta`). I think this is an acceptable tradeoff as this API is internal to core and I don't expect it to be widely used, but open to other opinions on it.

Reply:
Also, now that we have this, we could probably add the APM trace IDs using this API instead of hacking them directly into the logger as we have currently done: 
https://github.com/elastic/kibana/blob/6f9d771714d31a9bff2ae382e69bddd7bb2ef98e/packages/core/logging/core-logging-server-internal/src/logger.ts#L91-L97

We could do this by changing the signature of `GlobalMeta` to `Map<string, () => unknown>` so that the latest trace IDs could be retrieved on each entry, and setting the meta in `Root` so we can be sure it's there before core starts logging anything. I didn't do that in this PR to keep the scope as focused as possible, but can circle back and do a small follow-up if folks think it makes sense.

Reply:
> It would feel more intuitive to have a LogMeta object provided here

> it also means we are sacrificing the type safety of our ECS types

I ended up working around both of these concerns with a refactor. The way it works now is:
- When new context is added, loggingSystem stores it in a POJO. That means we can accept regular `LogMeta` after all, and it makes it easier to handle conflicts (the most recent write always "wins").
- Before updating each of the loggingAdapters, the loggingSystem flattens the object, so the loggingAdapter can use the flattened object entries to handle merging in the meta in a more performant way